modules/core/src/smithy4s/internals/PatternSegment.scala

```diff
@@ -0,0 +1,72 @@
+package smithy4s.internals
+
+import annotation.tailrec
+
+sealed abstract class PatternSegment(value: String)
+    extends Product
+    with Serializable {
+  def append(char: Char): PatternSegment
+}
+object PatternSegment {
+  final case class StaticSegment(val value: String)
+      extends PatternSegment(value) {
+    override def append(char: Char): PatternSegment =
+      this.copy(value = this.value + char)
+  }
+  final case class ParameterSegment(
+      val value: String,
+      terminationChar: Option[Char]
+  ) extends PatternSegment(value) {
+    override def append(char: Char): PatternSegment =
+      this.copy(value = this.value + char)
+  }
+
+  def segmentsFromString(str: String): List[PatternSegment] = {
```

Comment:
honestly, this would likely read better AND perform better using imperative style : even if it's not gonna run on the "hotpath", you need to think about the fact that we do cater to usecases like CLIs and ephemeral scripts, and deconstructing and reconstructing strings like. you're doing here is wildly inefficient memory wise ... 

 Like, I'd work with an index and iterate over the bytes of the array backing the String, and use a mutable `StringBuilder` to aggregate values.